keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1504 @@
+    // additional finalizers might also keep it around longer.
+    ResourceClaimFreed ResourceClaimPhase = “Freed”
+)
+
+type PodSpec {
+   ...
+   Resources []PodResource
+   ...
+}
+
+type Container {
+   ...
+   // The entries are the names of resources in PodSpec.Resources
+   // that are used by the container.
+   PodResources []string
+   ...
+}
+
+type PodResource struct {
+   // A name under which this resource can be referenced by the containers.
+   Name string
+
+   // The resource is independent of the Pod and defined by
+   // a separate ResourceClaim in the same namespace as
+   // the Pod. Either this or Template must be set, but not both.
+   ResourceClaimName *string
+
+    // Will be used to create a stand-alone ResourceClaim to allocate the resource.
+    // The pod in which this PodResource is embedded will be the
+    // owner of the ResourceClaim, i.e. the ResourceClaim will be deleted together with the
+    // pod.  The name of the ResourceClaim will be `<pod name>-<resource name>` where
+    // `<resource name>` is the name PodResource.Name
+    // Pod validation will reject the pod if the concatenated name
+    // is not valid for a ResourceClaim (for example, too long).
+    //
+    // An existing ResourceClaim with that name that is not owned by the pod
+    // will *not* be used for the pod to avoid using an unrelated
+    // resource by mistake. Starting the pod is then blocked until
+    // the unrelated ResourceClaim is removed. If such a pre-created ResourceClaim is
+    // meant to be used by the pod, the ResourceClaim has to be updated with an
+    // owner reference to the pod once the pod exists. Normally
+    // this should not be necessary, but it may be useful when
+    // manually reconstructing a broken cluster.
+    //
+    // This field is read-only and no changes will be made by Kubernetes
+    // to the ResourceClaim after it has been created.
+   Template *ResourceClaimTemplate
+}
+
+type ResourceClaimTemplate struct {
+    // May contain labels and annotations that will be copied into the ResourceClaim
+    // when creating it. No other fields are allowed and will be rejected during
+    // validation.
+    metav1.ObjectMeta
+
+    // The specification for the ResourceClaim. The entire content is
+    // copied unchanged into the ResourceClaim that gets created from this
+    // template. The same fields as in a ResourceClaim
+    // are also valid here.
+    Spec ResourceClaimSpec
+}
+```
+
+### Communication between kubelet and resource node plugin
+
+This gRPC interface is provided by the resource node plugin and invoked by
+kubelet. It is inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts removed.
+
+```
+<<[UNRESOLVED @pohly]>>
+Do plugin operations need secrets? They are currently not part of the proposed Kubernetes API.
+<<[/UNRESOLVED]>>
+```
+
+#### `NodePrepareResource`
```

Comment:
Why not call it `NodeAllocateResource` or `AllocateResources` as I think it better represents the functionality of this RPC? `NodePrepareResource` gives the impression that we are preparing the node but the actual allocation happens at a later (in another RPC) which is not the case here.

Reply:
No, the resource is already allocated before kubelet and thus this call here get involved.

Reply:
Are you referring to the case where devices are shared between various pods i.e. a device is already allocated to a pod but still has bandwidth to accommodate resource request of another pod and then we need to "prepare" the device for the other pod? 
If that is the case  it boils down when we consider a resource to have been allocated. IMO, it still makes sense to refer to the API as `AllocateResourcesas` as for that particular container/pod, the kubelet needs to be perform some action to make sure that the device can be **"allocated"** to the to that new pod.

Reply:
Our goal was to encourage plugins to do as much of the "allocation" as possible before a pod runs. Then `NodePrepareResource` should be so light-weight that it cannot fail or at least that it is extremely unlikey. The motivation for that approach is that there is no way how a pod can be moved to a different node if `NodePrepareResource` fails because, by definition, it only gets called after a pod has been permanently scheduled unto a node.

Plugins can deviate from that. They can do nothing when setting "phase: allocated" and do all of the actual allocation during `NodePrepareResource`. But that's not how it should be done.